aten/src/ATen/xpu/CachingHostAllocator.h

```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <ATen/core/CachingHostAllocator.h>
+#include <ATen/xpu/XPUEvent.h>
+#include <c10/core/Allocator.h>
+#include <c10/xpu/XPUStream.h>
+
+namespace at::xpu {
```

Comment:
Any reason why we can't have these APIs as generic accelerator APIs?

Reply:
Yes, we can have generic APIs to replace these C-like APIs.
We are going to submit another PR to unify them, to avoid many cuda modifications in this PR.
Is it OK to you? @albanD

Reply:
These APIs are aligned to CUDA, see https://github.com/pytorch/pytorch/blob/a4c8002ee0714db7d19d98dc0c84108521c5ce6f/aten/src/ATen/cuda/CachingHostAllocator.h#L21-L35

These APIs `record_event`, `empty_cache`, and `allocate` are already defined in [CachingHostAllocatorInterface](https://github.com/pytorch/pytorch/blob/a4c8002ee0714db7d19d98dc0c84108521c5ce6f/aten/src/ATen/core/CachingHostAllocator.h#L352-L378). As similar as device caching allocator, `getCachingHostAllocator` returns a `c10::Allocator*` that doesn't have `record_event` and `empty_cache`.
Do you mean why don't we define `record_event` and `empty_cache` in `c10::Allocator`? If we do this, we can directly can `at::xpu::getCachingHostAllocator().empty_cache()` or `at::cuda::getCachingHostAllocator().record_event()` instead of `at::xpu::CachingHostAllocator_emptyCache()` or `at::cuda::CachingHostAllocator_recordEvent()`, right? Even further, referring to the device allocator, we can call `at::getHostAllocator(device_type).empty_cache()` as a generic method more conveniently.

I think it is a good idea. And it is our final goal. But currently, we plan to implement it step by step. In this PR, we mainly propose to generalize host caching allocator as device-agnostic, according to [[RFC] Intel GPU Runtime Upstreaming for Allocator](https://github.com/pytorch/pytorch/issues/116322), and add support of XPU host allocator sharing the same caching mechanism and code with CUDA.
 - Next, we aim to generalize device allocator (share the code across each backend).
 - Finally, we intend to unify device and host allocator (have `empty_cache` as generic API in `c10::Allocator`).

Meanwhile, there are some open need to be discussed, like:
 - there are at least two types of allocator in PyTorch: caching, and not caching(like async device allocator). Apparently, `empty_cache` should be the caching allocator's method rather than not the caching allocator. Is it reasonable to define `empty_cache` in `c10::Allocator`? Does it seem like `c10::Allocator` is always a caching allocator?
 - record_event(named in host allocator) accepts three parameters as its inputs since host allocator does not split larger allocations into smaller blocks, unlike the caching device allocator. However, record_stream(named in device allocator) accepts two parameters as its inputs. Do we need to unify them or leave them as they are?

In summary, we leave these APIs here aligning to CUDA and make sure this PR change is relatively small. We accept these APIs as generic accelerator APIs if I capture your point. We will unify them following our plan, 1) first generalize device allocator, and then 2) unify them

Reply:
Do you mean to rename `at::xpu::CachingHostAllocator_emptyCache` to `at::xpu::CachingHostAllocator::empty_cache` including CUDA?

Reply:
I think the idea is `at::getHostAllocator(device_type).empty_cache()` yes.
With on top of that, `device_type` being optional and defaulting to the current accelerator `at::getAccelerator()` so that most callers don't have to specify the device_type anyways.

And yes we should use `c10::Stream` to handle the stream objects.

Reply:
Got your point.